File: Darts/src/burlton/dartzee/code/screen/stats/player/PlayerAchievementsScreen.java
package burlton.dartzee.code.screen.stats.player;

import burlton.dartzee.code.achievements.AchievementConstants;
import burlton.dartzee.code.bean.AchievementMedal;
import burlton.dartzee.code.db.AchievementEntity;
import burlton.dartzee.code.db.PlayerEntity;
import burlton.dartzee.code.screen.EmbeddedScreen;
import burlton.dartzee.code.screen.ScreenCache;
import javafx.scene.paint.Color;

import javax.swing.*;
import java.awt.*;

public final class PlayerAchievementsScreen extends EmbeddedScreen
{
	private PlayerEntity player = null;
	
	
	public PlayerAchievementsScreen() 
	{	
		JTabbedPane tabbedPane = new JTabbedPane(SwingConstants.TOP);
		add(tabbedPane, BorderLayout.CENTER);
		
		
		tabbedPane.addTab("General", null, panelGeneral, null);
		
		panelGeneral.add(gray);
		panelGeneral.add(red);
		panelGeneral.add(orange);
		panelGeneral.add(yellow);
		panelGeneral.add(green);
		panelGeneral.add(cyan);
		panelGeneral.add(hotpink);
		
		
		//FlowLayout flowLayout = (FlowLayout) panelX01.getLayout();
		//flowLayout.setVgap(20);
		//flowLayout.setHgap(20);
		//flowLayout.setAlignment(FlowLayout.LEFT);
		//tabbedPane.addTab("X01", null, panelX01, null);
			
		//panelX01.add(lblNewLabel);
	}
	
	private final JPanel panelGeneral = new JPanel();
	private final AchievementMedal gray = new AchievementMedal(15, Color.GRAY);
	private final AchievementMedal red = new AchievementMedal(45, Color.RED);
	private final AchievementMedal orange = new AchievementMedal(85, Color.ORANGE);
	private final AchievementMedal yellow = new AchievementMedal(180, Color.YELLOW);
	private final AchievementMedal green = new AchievementMedal(210, Color.LIGHTGREEN);
	private final AchievementMedal cyan = new AchievementMedal(266, Color.CYAN);
	private final AchievementMedal hotpink = new AchievementMedal(360, Color.DEEPPINK);
	private final JLabel lblNewLabel = new JLabel("");
	
	@Override
	public String getScreenName()
	{
		return "Achievements - " + player.getName();
	}

	@Override
	public void initialise()
	{
		long playerId = player.getRowId();
		
		AchievementEntity achievement = AchievementEntity.retrieveAchievement(AchievementConstants.ACHIEVEMENT_REF_X01_BEST_FINISH, playerId);
		if (achievement != null)
		{
			lblNewLabel.setText("Best Finish: " + achievement.getAchievementCounter() + ", Game #" + achievement.getGameIdEarned());
		}
		else
		{
			lblNewLabel.setText("");
		}
	}
	
	/*@Override
	public void postInit()
	{
		HandyArrayList<AchievementMedal> medals = ComponentUtil.getAllChildComponentsForType(this, AchievementMedal.class);
		for (AchievementMedal medal : medals)
		{
			medal.animateProgressBar();
		}
	}*/
	
	@Override
	public EmbeddedScreen getBackTarget()
	{
		return ScreenCache.getPlayerManagementScreen();
	}

	public void setPlayer(PlayerEntity player)
	{
		this.player = player;
	}
}

File: Darts/src/burlton/dartzee/code/screen/stats/player/PlayerAchievementsScreen.kt
package burlton.dartzee.code.screen.stats.player

import burlton.dartzee.code.achievements.AbstractAchievement
import burlton.dartzee.code.achievements.getAchievementMaximum
import burlton.dartzee.code.achievements.getAchievementsForGameType
import burlton.dartzee.code.achievements.getPlayerAchievementScore
import burlton.dartzee.code.bean.AchievementMedal
import burlton.dartzee.code.db.AchievementEntity
import burlton.dartzee.code.db.PlayerEntity
import burlton.dartzee.code.screen.EmbeddedScreen
import burlton.dartzee.code.screen.ScreenCache
import burlton.dartzee.code.utils.getAllGameTypes
import burlton.dartzee.code.utils.getTypeDesc
import burlton.dartzee.code.core.bean.WrapLayout
import java.awt.*
import javax.swing.*
import javax.swing.border.EmptyBorder

class PlayerAchievementsScreen : EmbeddedScreen()
{
    var previousScrn: EmbeddedScreen = ScreenCache.getPlayerManagementScreen()
    private var player: PlayerEntity? = null
    private var progressDesc = ""

    private val centerPanel = JPanel()
    private val tabbedPane = JTabbedPane(SwingConstants.TOP)
    private val panelAchievementDesc = JPanel()
    private val lblAchievementName = JLabel()
    private val lblAchievementDesc = JLabel()
    private val lblAchievementExtraDetails = JLabel()

    init
    {
        add(centerPanel, BorderLayout.CENTER)

        centerPanel.layout = BorderLayout()
        centerPanel.add(tabbedPane, BorderLayout.CENTER)

        centerPanel.add(panelAchievementDesc, BorderLayout.SOUTH)
        panelAchievementDesc.preferredSize = Dimension(200, 100)
        panelAchievementDesc.layout = BorderLayout()

        panelAchievementDesc.add(lblAchievementName, BorderLayout.NORTH)
        lblAchievementName.horizontalAlignment = JLabel.CENTER
        lblAchievementName.font = Font("Trebuchet MS", Font.BOLD, 24)
        lblAchievementDesc.horizontalAlignment = JLabel.CENTER
        lblAchievementDesc.font = Font("Trebuchet MS", Font.PLAIN, 20)
        lblAchievementExtraDetails.horizontalAlignment = JLabel.CENTER
        lblAchievementExtraDetails.font = Font("Trebuchet MS", Font.ITALIC, 18)

        panelAchievementDesc.add(lblAchievementDesc, BorderLayout.CENTER)
        panelAchievementDesc.add(lblAchievementExtraDetails, BorderLayout.SOUTH)

        panelAchievementDesc.border = EmptyBorder(5, 5, 5, 5)
    }

    override fun getScreenName() : String
    {
        return "Achievements - ${player?.name} - $progressDesc"
    }


    override fun initialise()
    {
        tabbedPane.removeAll()

        val playerId = player?.rowId!!

        val achievementRows = AchievementEntity.retrieveAchievements(playerId)
        getAllGameTypes().forEach {
            addAchievementTab(it, achievementRows)
        }

        val score = getPlayerAchievementScore(achievementRows, player!!)
        progressDesc = "$score/${getAchievementMaximum()}"
    }

    private fun addAchievementTab(gameType: Int, achievementRows: List<AchievementEntity>)
    {
        val panel = JPanel()

        val fl = WrapLayout()
        fl.vgap = 25
        fl.hgap = 20
        fl.alignment = FlowLayout.LEFT
        panel.layout = fl

        val sp = JScrollPane()
        sp.setViewportView(panel)
        sp.horizontalScrollBarPolicy = ScrollPaneConstants.HORIZONTAL_SCROLLBAR_NEVER
        sp.verticalScrollBar.unitIncrement = 16

        tabbedPane.addTab(getTypeDesc(gameType), sp)

        getAchievementsForGameType(gameType).forEach {
            addAchievement(it, achievementRows, panel)
        }
    }

    private fun addAchievement(aa: AbstractAchievement, achievementRows: List<AchievementEntity>, panel: JPanel)
    {
        val ref = aa.achievementRef
        val achievementRowsFiltered = achievementRows.filter { a -> a.achievementRef == ref }.toMutableList()

        aa.initialiseFromDb(achievementRowsFiltered, player)

        val medal = AchievementMedal(aa)
        panel.add(medal)
    }

    fun toggleAchievementDesc(hovered: Boolean, achievement : AbstractAchievement)
    {
        if (hovered)
        {
            val color = achievement.getColor(false)
            setPanelColors(color, color.darker().darker())

            lblAchievementName.text = achievement.name

            if (!achievement.isLocked())
            {
                lblAchievementDesc.text = achievement.desc
                lblAchievementExtraDetails.text = achievement.getExtraDetails()
            }
        }
        else
        {
            setPanelColors(null, null)
            lblAchievementName.text = ""
            lblAchievementDesc.text = ""
            lblAchievementExtraDetails.text = ""
        }
    }

    private fun setPanelColors(bgColor: Color?, fgColor: Color?)
    {
        panelNavigation.background = bgColor
        panelAchievementDesc.background = bgColor
        panelNext.background = bgColor
        panelBack.background = bgColor

        lblAchievementName.background = bgColor
        lblAchievementDesc.background = bgColor
        lblAchievementExtraDetails.background = bgColor

        lblAchievementName.foreground = fgColor
        lblAchievementDesc.foreground = fgColor
        lblAchievementExtraDetails.foreground = fgColor
    }


    override fun getBackTarget(): EmbeddedScreen
    {
        return previousScrn
    }

    fun setPlayer(player: PlayerEntity)
    {
        this.player = player
    }

    override fun getDesiredSize(): Dimension?
    {
        return Dimension(1240, 700)
    }
}
